ee/tabby-db/src/integration_provider.rs

```diff
@@ -0,0 +1,258 @@
+        query!(
+            "UPDATE integration_access_tokens SET updated_at = ?, access_token = IIF(?, access_token, NULL) WHERE id = ?",
+            time,
+            success,
+            id
+        )
+        .execute(&self.pool)
+        .await?;
+        Ok(())
+    }
+
+    pub async fn list_integration_access_tokens(
+        &self,
+        ids: Vec<i64>,
+        limit: Option<usize>,
+        skip_id: Option<i32>,
+        backwards: bool,
+    ) -> Result<Vec<IntegrationAccessTokenDAO>> {
+        let condition = (!ids.is_empty()).then(|| {
+            let ids = ids
+                .into_iter()
+                .map(|id| id.to_string())
+                .collect::<Vec<_>>()
+                .join(", ");
+            format!("id in ({ids})")
+        });
+        let providers = query_paged_as!(
+            IntegrationAccessTokenDAO,
+            "integration_access_tokens",
+            [
+                "id",
+                "kind",
+                "display_name",
+                "access_token",
+                "created_at" as "created_at: DateTimeUtc",
+                "updated_at" as "updated_at: DateTimeUtc"
+            ],
+            limit,
+            skip_id,
+            backwards,
+            condition
+        )
+        .fetch_all(&self.pool)
+        .await?;
+        Ok(providers)
+    }
+
+    pub async fn upsert_provided_repository(
+        &self,
+        integration_access_token_id: i64,
+        vendor_id: String,
+        name: String,
+        git_url: String,
+    ) -> Result<i64> {
+        let res = query!(
+            "INSERT INTO provided_repositories (access_token_provider_id, vendor_id, name, git_url) VALUES ($1, $2, $3, $4)
+                ON CONFLICT(access_token_provider_id, vendor_id) DO UPDATE SET name = $3, git_url = $4, updated_at = DATETIME('now')",
+            integration_access_token_id,
+            vendor_id,
+            name,
+            git_url
+        ).execute(&self.pool).await?;
+        Ok(res.last_insert_rowid())
+    }
+
+    pub async fn delete_outdated_provided_repositories(
+        &self,
+        integration_access_token_id: i64,
+        cutoff_timestamp: DateTimeUtc,
+    ) -> Result<usize> {
+        let res = query!(
+            "DELETE FROM provided_repositories WHERE access_token_provider_id = ? AND updated_at < ?;",
+            integration_access_token_id,
+            cutoff_timestamp
+        ).execute(&self.pool).await?;
+        Ok(res.rows_affected() as usize)
+    }
+
+    pub async fn get_provided_repository(&self, id: i64) -> Result<ProvidedRepositoryDAO> {
+        let repo = query_as!(
+            ProvidedRepositoryDAO,
+            "SELECT id, vendor_id, name, git_url, active, access_token_provider_id, created_at, updated_at FROM provided_repositories WHERE id = ?",
+            id
+        )
+        .fetch_one(&self.pool)
+        .await?;
+        Ok(repo)
+    }
+
+    pub async fn list_provided_repositories(
```

Comment:
@wsxiaoys Should this have an additional filter for `kind`? You can still effectively do that by providing all the IDs for a given kind, but this might make it easier to use